Add a basic implementation of ReadableStream async iterable

youennf opened the pull request:
#### eb724e8b508bfd8d17bd76d33ce5d3992b292274
<pre>
Add a basic implementation of ReadableStream async iterable
<a href="https://rdar.apple.com/165605714">rdar://165605714</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=303302">https://bugs.webkit.org/show_bug.cgi?id=303302</a>

Reviewed by Chris Dumez.

We add a basic ReadableStream Iterator implementation.
As per <a href="https://streams.spec.whatwg.org/#rs-asynciterator">https://streams.spec.whatwg.org/#rs-asynciterator</a>, we implement the init steps and the get next iteration result steps.
A follow-up will implement the return steps.

This is implemented in ReadableStream::Iterator class which takes a reader from its stream.
Given it might throw, we add support for throwing in iteratorCreate.

To do so, we create the internal iterator and pass it to JSDOMIteratorBase or JSDOMAsyncIteratorBase.
We add support for passing parameters to the iterator creation steps and make use of it for passing ReadableStreamIteratorOptions.

We do not yet support converting iterator creation parameters, the binding generator only creates default values (in this case preventCancel = false). This will be fixed in a follow-up.

We also update the binding generator to not expose the entries and keys methods for async iterables that are not key value.
We do not do so for sync iterator since that would break FontFaceSet (which should be moved to setlike).

We update the binding generator according the new iteratorCreate function.

Covered by rebased WPT tests and rebased binding tests.

Canonical link: <a href="https://commits.webkit.org/303734@main">https://commits.webkit.org/303734@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/606e1c3532df6e9e00b39d3f443b2403078b8d92

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133485 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5986 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44630 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141041 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85534 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135355 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6501 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5851 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/102112 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85534 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136432 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4603 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/119661 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/82909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/4481 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/2082 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/113601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/37771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143688 "Built successfully") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/5656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/38356 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/110488 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/5738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/4850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/110670 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/115918 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/59406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20638 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/5711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/34234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/5557 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69163 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/5800 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/5667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

stwrt commented on LayoutTests/imported/w3c/web-platform-tests/compression/compression-including-empty-chunk.any-expected.txt:
```diff
@@ -8,7 +8,7 @@ PASS the result of compressing [Hello,Hello,] with deflate-raw should be 'HelloH
-FAIL the result of compressing [,Hello,Hello] with brotli should be 'HelloHello' assert_array_equals: value should match lengths differ, expected array object "" length 0, got object "72,101,108,108,111,72,101,108,108,111" length 10
-FAIL the result of compressing [Hello,,Hello] with brotli should be 'HelloHello' assert_array_equals: value should match lengths differ, expected array object "" length 0, got object "72,101,108,108,111,72,101,108,108,111" length 10
-FAIL the result of compressing [Hello,Hello,] with brotli should be 'HelloHello' assert_array_equals: value should match lengths differ, expected array object "" length 0, got object "72,101,108,108,111,72,101,108,108,111" length 10
+PASS the result of compressing [,Hello,Hello] with brotli should be 'HelloHello'
```
Great to see the brotli tests passing again with this patch.

youennf committed:
Add a basic implementation of ReadableStream async iterable
rdar://165605714
https://bugs.webkit.org/show_bug.cgi?id=303302

Reviewed by Chris Dumez.

We add a basic ReadableStream Iterator implementation.
As per https://streams.spec.whatwg.org/#rs-asynciterator, we implement the init steps and the get next iteration result steps.
A follow-up will implement the return steps.

This is implemented in ReadableStream::Iterator class which takes a reader from its stream.
Given it might throw, we add support for throwing in iteratorCreate.

To do so, we create the internal iterator and pass it to JSDOMIteratorBase or JSDOMAsyncIteratorBase.
We add support for passing parameters to the iterator creation steps and make use of it for passing ReadableStreamIteratorOptions.

We do not yet support converting iterator creation parameters, the binding generator only creates default values (in this case preventCancel = false). This will be fixed in a follow-up.

We also update the binding generator to not expose the entries and keys methods for async iterables that are not key value.
We do not do so for sync iterator since that would break FontFaceSet (which should be moved to setlike).

We update the binding generator according the new iteratorCreate function.

Covered by rebased WPT tests and rebased binding tests.

Canonical link: https://commits.webkit.org/303734@main
